### PR TITLE
Add support for additional languages

### PR DESCRIPTION
**Pull request by elyas-bhy**
This PR resolves issues #104 in a backwards-compatible way.
It adds the ability to specify the language of the input text to analyze.

It relies on some changes introduced in PR #93 such as relying on the [`afinn-165-multilingual` repository](https://github.com/dkocich/afinn-165-multilingual) to retrieve AFINN-165 translations for over 100+ languages.

Although the accuracy of these translations is yet to be determined (as discussed in [this comment](https://github.com/thisandagain/sentiment/issues/104#issue-206686881)), this change adds the ability to easily swap the translations if needed, with minimal refactoring.
It also enables developers (with lower requirements wrt. precision) to access these translations directly from this module, instead of having to fork the project (as in the comments of issue #104, see https://github.com/AlexGustafsson/sentiment-swedish and https://github.com/kubawolanin/sentiment-polish for example).

Finally, this PR also introduces the ability to define language-specific negation strategies, as negation can be more or less complex according to the language, and does not consist in simply translating negators from one language to another.

### CODE REVIEW

**Comment by thisandagain** [comment]
@elyas-bhy Thanks! I'll take a look at this over the next week. 😄

**Comment by elyas-bhy** [comment]
@thisandagain have you had the time to review this PR?

**Comment by thisandagain** [comment]
Thanks @elyas-bhy! I think this is an interesting and exciting approach. Couple comments and questions that I'd love to get your thoughts on:

1. This feels like such a fundamental change that I'd be ok to make this breaking and bump the major version number. In some places you are jumping through hoops to maintain compatibility (such as [here](https://github.com/thisandagain/sentiment/pull/128/files#diff-6d186b954a58d5bb740f73d84fe39073R16)) – which I greatly appreciate – but I think I'd be ok to rethink the API to make sure things stay simple.

2. In general I don't feel great about bringing in the translations from https://github.com/dkocich/afinn-165-multilingual. I don't speak that many languages, but from the few that I do I can tell that the quality of some of the translations provided by Google Translate is quite poor. I think I'd rather think about a structure where contributors can submit a single "package" that covers the AFINN translation dictionary, negation works, and a labeled dataset for validation. In that way perhaps we could rethink that the existing dictionary, negation list, and UCI datasets to be a default "en" package.

3. Totally a detail / nitpick: I suppose I feel about "helper" the same way I do as having a module called "util". I think I'd prefer that any modules added have names that make it immediately clear what they do.

4. I would want to make sure that any new code landed maintains the current level (100%) of test coverage. You can run the coverage report using `make coverage`. It's not too bad ... just a few lines in the helper that aren't covered.

**Comment by elyas-bhy** [comment]
> 1. This feels like such a fundamental change that I'd be ok to make this breaking and bump the major version number. In some places, you are jumping through hoops to maintain compatibility (such as here) – which I greatly appreciate – but I think I'd be ok to rethink the API to help things stay simple.

Sure thing. What kind of API do you have in mind?
I was thinking that the most flexible API would be using setters, like in object-oriented programming:

```js
const Sentiment = require('sentiment');  // a constructor is exported
const sentiment = new Sentiment();  // create a new instance
sentiment.setLanguage('en');
sentiment.setExtras({ ... });  // inject custom keywords
sentiment.addExtras({ ... });  // extend the existing keywords
sentiment.ignoreEmojis(true);  // ignore emojis
// etc. for other options
const result = sentiment.analyze('Hello world');
```

This has the benefit of being flexible enough to allow adding further options without breaking the existing API. Thoughts?


> 2. In general, I don't feel great about bringing in the translations from https://github.com/dkocich/afinn-165-multilingual. I don't speak many languages, but from the few that I do, I can tell that some of the translations provided by Google Translate are quite poor. I think I'd rather think about a structure where contributors can submit a single "package" that covers the AFINN translation dictionary, negation works, and a labeled dataset for validation. In that way perhaps we could rethink that the existing dictionary, negation list, and UCI datasets to be a default (en) "package".

I agree that the translations from Google Translate are quite lacking, but at least they allow developers to start testing them in order to iterate and refine them later.
A good compromise for this PR is to include the necessary refactoring for supporting additional languages, but only expose english for now, since it is the only language that is thoroughly tested. What are your thoughts on this?
Also, there was a discussion in issue #104 about using CrowdFlower for establishing labeled corpora for different languages. Is this something that is still planned? What can we do to push this forward?


> 3. Totally a detail / nitpick: I suppose I feel about "helper" the same way I do as having a module called "util". I think I'd prefer that any modules added have names that make it immediately clear what they do.

Agreed. I'll fix this once we agree on the more important changes mentioned above.


> 4. Also a detail: I would want to make sure that any new code landed maintains the current level (100%) of test coverage. You can run the coverage report using make coverage. It's not too bad ... just a few lines in the helper that aren't covered!

Same as above.

**Comment by elyas-bhy** [comment]
@thisandagain thoughts?

**Comment by thisandagain** [comment]
@elyas-bhy Thanks for being so thoughtful about this.

Using an OOP setter paradigm seems reasonable to me and could help keep the implementation clean. I can see an argument for using a configuration object paradigm (which feels more javascript-y) and less verbose:
```js
const sentiment = require('sentiment');
const result = sentiment({
    text: 'Hello world',  // required
    extras: [],           // optional
    language: 'en'        // optional
    emoji: true .         // optional
})
```

Performance-wise it's a little hard to say what the impact of either approach would be. I can see pros and cons to each.

---

> "A good compromise for this PR is to include the necessary refactoring for supporting additional languages, but only expose english for now, since it is the only language that is thoroughly tested. What are your thoughts on this?" 

👍. That feels like a perfectly reasonable compromise. Thanks!

**Comment by elyas-bhy** [comment]
@thisandagain I like your suggestion. However, I feel like since the `text` is always required, it should be its own argument, separate from the optional ones, like so:

```javascript
const sentiment = require('sentiment');
const result = sentiment('Hello world', {
    extras: [],         // optional
    language: 'en'      // optional
    emoji: true .       // optional
})
```

This feels more sensible to me, and is in line with common Javascript practices.
Thoughts?

**Comment by thisandagain** [comment]
@elyas-bhy Great point. That looks good to me! 👍

**Comment by elyas-bhy** [comment]
@thisandagain Regarding returning the result, do we keep the callbacks? In which case, would something like `sentiment(phrase, opts, callback)` make sense? Or should we return promises? Or should we support both paradigms (i.e. return a promise if no callback was specified)?

**Comment by elyas-bhy** [comment]
@thisandagain I finished integrating the changes we discussed. Please check out the updated README file as it summarizes the new features and modifications. Note that merging this PR would require releasing a new major version as it has breaking changes.

**Comment by thisandagain** [comment]
@elyas-bhy Related: I'm going to add you as a contributor to this repo. It's the very least I can do to recognize how great your contributions have been. 😄

**Comment by elyas-bhy** [comment]
Thanks for the feedback. I will look into this when I get back on Monday.
Also thanks for the contributor rights.

**Comment by elyas-bhy** [comment]
> This would lead me to suspect that the "getLabels" lookup for the language processor may be the culprit. Luckily, I think this may be easy to resolve by simply making sure that the english labels are loaded at "require-time" rather than every time that the sentiment.analyze function is run. In addition, it also points to a potential benefit of holding the labels in a cache object (particularly when users are leveraging something other than english).

This is already the case:

- [English language is loaded at require-time and cached](https://github.com/elyas-bhy/sentiment/blob/develop/lib/language-processor.js#L3-L5)
- [The `getLabels` function calls the `getLanguage` function which caches languages when they are first loaded](https://github.com/elyas-bhy/sentiment/blob/develop/lib/language-processor.js#L27-L44)

After some digging into the profiler, it turns out that the performance hit is due to these lines:
```javascript
// https://github.com/elyas-bhy/sentiment/blob/develop/lib/index.js#L47-L49
if (opts.emojis !== false) {
  Object.assign(labels, emojis);
}
```
In other words, providing the ability to enable or disable emoji detection as an option of the `sentiment.analyze` function leads to a major performance hit.
We should probably rethink this option, is it really necessary after all? Perhaps we can enable emoji detection by default, and eventually, if needed, provide an option to disable it at a global level, to avoid merging large objects at runtime.

@thisandagain thoughts?

**Comment by elyas-bhy** [comment]
@thisandagain ping.

**Comment by elyas-bhy** [comment]
@thisandagain ping again.

**Comment by thisandagain** [comment]
Sorry for the delay @elyas-bhy. I see ... that all makes sense. The issue is that the emoji processing is still being applied even if `emojis` is undefined in `opts`. By default we omit the emoji processing due to the performance implications (discussion [here](https://github.com/thisandagain/sentiment/pull/74#issuecomment-289233208)). I think swapping out that logic with the following should resolve this issue:

```javascript
// Add emojis unless explicitly excluded
if (typeof opts.emojis !== 'undefined') {
     if (opts.emojis) Object.assign(labels, emojis);
}
```

Resulting benchmark:
```
sentiment (Latest) - Short  x 867,995 ops/sec ±2.02% (89 runs sampled)
sentiment (Latest) - Long   x 3,793 ops/sec ±1.45% (88 runs sampled)
```

**Comment by elyas-bhy** [comment]
@thisandagain thanks for the feedback. However, I think I have not properly explained what I had in mind in my last comment.

Sure, if you disable emojis by default, then the benchmarks seem to perform way better, but I find that a bit misleading. The fact is that, even with your proposed fix, we are still calling `Object.assign` every time we `analyze` something. This is inefficient, because you are rebuilding everything at each subsequent `analyze` call.

My proposed fix is to merge the emoji labels *once*, when initializing the `sentiment` instance, to avoid repeatedly merging large objects at runtime. If we go this route, we need to ask ourselves if we need a way to disable emojis if needed (does the community really need this option after all?). Perhaps requiring the `sentiment` module should return a constructor, so that the user can pass options to create an instance? This approach is useful and future-proof to enable to pass global options, and should be considered since we are releasing a new major version anyway.

Thoughts?

**Comment by thisandagain** [comment]
@elyas-bhy Ah! Thanks for clarifying. That makes sense. My hesitation with landing the emoji work initially is that it would be a breaking change and we had poor validation tests at the time. Since this will be a breaking change anyway I think I'm ok to land it as long as the overall performance and validation tests look ok.

> Perhaps requiring the sentiment module should return a constructor, so that the user can pass options when creating a new instance? This approach is useful and future-proof to enable passing global options, and should be considered since we are releasing a new major version anyway.

That seems reasonable and should provide some users piece of mind until we have more robust validation available for the emoji sentiment.

---

I think this is really close. Thanks for working through this with me @elyas-bhy.

**Comment by elyas-bhy** [comment]
@thisandagain ping.

**Comment by thisandagain** [comment]
... and merged! Thanks a ton for all your work on this @elyas-bhy. Also thank you for being so patient with me and my crazy schedule this year. 💟

**Comment by dkocich** [comment]
@elyas-bhy thanks for your work on this issue :-)
